book/src/move-basics/enum-and-match.md

```diff
@@ -0,0 +1,287 @@
+
+```move
+/// Returns `Some(String)` if the `Segment` is `String`, `None` otherwise.
+public fun try_into_inner_string(s: Segment): Option<String> {
+    match (s) {
+        Segment::String(str) => option::some(str),
+        _ => option::none()
+    }
+}
+```
+
+This pattern safely exposes internal data in a controlled way, avoiding abort.
+
+### Trick #3 - Matching on primitive values
+
+The `match` expression in Move can be used with values of any type – enums, structs, or primitives.
+To demonstrate this, let’s implement a `to_string` function that creates a new
+`String from a `Segment`. In the case of the `Special`variant, we will match on the`encoding` field
+to determine how to decode the content.
+
+```move
+/// Return a `String` representation of a segment.
+public fun to_string(s: &Segment): String {
+    match (*s) {
+        // Return an empty string.
+        Segment::Empty => b"".to_string(),
+        // Return the inner string.
+        Segment::String(str) => str,
+        // Return the decoded contents based on the encoding.
+        Segment::Special { content, encoding } => {
+            // perform a match on the encoding, we only support 0 - utf8, 1 - hex
+            match (encoding) {
+                // Plain encoding, return content
+                0 => content.to_string(),
+                // HEX encoding, decode and return
+                1 => sui::hex::decode(content).to_string(),
+                // We have to provide a wildcard pattern, because values of `u8` are 0-255
+                _ => abort
+            }
+        }
+    }
+}
+```
+
+This function demonstrates two key things:
+
+- Nested `match` expressions can be used for deeper logic branching.
+- Wildcards are essential for covering all possible values in primitive types like `u8`.
+
+## The final test
+
+Now we can finalize the test we started before using the features we have added. Let's create a
+scenario where we build enums into a vector.
+
+```move
+// note, that the module has changed!
+module book::segment_tests;
+
+use book::segment;
+use std::unit_test::assert_eq;
+
+#[test]
+fun test_full_enum_cycle() {
+    // create a vector of different Segment variants
+    let segments = vector[
+        segment::new_empty(),
+        segment::new_string(b"hello".to_string()),
+        segment::new_special(b" ", 0), // plaintext
+        segment::new_string(b"move".to_string()),
+        segment::new_special(b"21", 1), // hex
+    ];
+
+    // aggregate all segments into the final string using `vector::fold!` macro
+    let result = segments.fold!(b"".to_string(), |mut acc, segment| {
+        // do not append empty, only `Special` and `String`
+        if (!segment.is_empty()) {
+            acc.append(segment.to_string());
+        };
+        acc
+    });
+
+    // check that the result
+    assert_eq!(result, b"hello move!".to_string());
+}
+```
+
+Hooray! This test demonstrates the full enum workflow: instantiating different variants, using
```

Comment:
```suggestion
This test demonstrates the full enum workflow: instantiating different variants, using
```
Full disclosure, I hate exclamation points! But the hooray here seems particularly out of place? Not sure what we're hooraying. Maybe add it to the end of the paragraph? Then you can have two! exclamations in a row!